lms/djangoapps/grades/tasks.py

```diff
@@ -115,6 +115,34 @@ def compute_grades_for_course(course_key, offset, batch_size, **kwargs):  # pyli
             raise result.error
 
 
+@task(
+    bind=True,
+    time_limit=SUBSECTION_GRADE_TIMEOUT_SECONDS,
+    max_retries=2,
+    default_retry_delay=RETRY_DELAY_SECONDS,
+    routing_key=settings.RECALCULATE_GRADES_ROUTING_KEY
+)
+def recalculate_course_and_subsection_grades_for_user(self, **kwargs):
+    """
+    Recalculates the course grade and all subsection grades
+    for the given ``user`` and ``course_key`` keyword arguments.
+    """
+    user = kwargs.get('user')
+    course_key = kwargs.get('course_key')
+
+    if not user or course_key:
+        message = 'recalculate_course_and_subsection_grades_for_user missing "user" or "course_key" kwargs from {}'
+        log.error(message.format(kwargs))
+
+    previous_course_grade = CourseGradeFactory().read(user, course_key=course_key)
+    if previous_course_grade and previous_course_grade.attempted:
+        CourseGradeFactory().update(
+            user=user,
+            course_key=course_key,
+            force_update_subsections=True,
+            collected_block_structure=previous_course_grade.course_data.collected_structure
```

Comment:
Given that `previous_course_grade` is read from disk and wouldn't have `collected_structure` populated, this line doesn't seem to add value.

Reply:
Hmm, on my devstack, there's something in that attribute:
```python
>>> from django.contrib.auth.models import User
>>> user = User.objects.get(username='staff')
>>> user
<User: staff>
>>> from opaque_keys.edx.keys import CourseKey
>>> key = CourseKey.from_string('course-v1:edX+DemoX+Demo_Course')
>>> from lms.djangoapps.grades.course_grade_factory import *
>>> grade = CourseGradeFactory().read(user, course_key=key)
>>> grade.course_data
<lms.djangoapps.grades.course_data.CourseData object at 0x7f973d484650>
>>> grade.course_data.collected_structure
2018-05-07 14:48:21,544 INFO 6644 [openedx.core.djangoapps.content.block_structure.store] store.py:177 - BlockStructure: Read from cache; <openedx.core.djangoapps.content.block_structure.store.StubModel object at 0x7f973be47bd0>, size: 30139
<openedx.core.djangoapps.content.block_structure.block_structure.BlockStructureBlockData object at 0x7f973bde3550>
>>> list(grade.course_data.collected_structure.iteritems())
# a list full of BlockUsageLocators...
```
...but I also don't have any PersistentCourseGrade records in my local DB - could that be why this is collecting block structure?

Reply:
I think it's because the persisted grades waffle flags aren't enabled on your devstack.
Have we updated devstacks yet to enable them by default?

Reply:
Ah, that's probably it.  I don't think we enable those by default yet.

Reply:
No worries.  Would you create an Educator ticket to do so?  @robrap may also have some tips on how to enable by default.

Reply:
I guess I'm also forcing the evaluation of `collected_structure` from the shell, thus populating that attribute on the course data object :)

When I look at the private attributes `_structure` and `_collected_block_structure`, they are empty.  Anyway, thanks for the help.

Reply:
@iloveagent57 @nasthagiri: We don't have a great story for updating defaults yet.  

The WaffleFlag class takes a default to use if it hasn't been set, which we have used, but it has its pros and cons.  The pro is that it gets the job done and it won't override/overwrite a value once it is set through admin.  The con is that it can be very confusing that a flag that doesn't exist in admin may implicitly be on due to code.  We may end up removing this.

Another way is through migrations, but we don't have great examples for this.  The code here should look to see if the flag already exists before updating it, because you don't want to force the flag on in an environment where someone explicitly set it to be false through the admin.  I haven't implemented this yet.

There is also a way to set a flag on by default for bokchoy tests if that is something that is needed.  You can search this page for "default" or "bokchoy" for more details: https://openedx.atlassian.net/wiki/spaces/OpenDev/pages/40862688/Feature+Flags+and+Settings+on+edx-platform

Unfortunately, I don't think there is a great best practice to recommend at this point, and maybe you can help improve and develop one, but this comment at least outlines some options.